ecl/hthor/hthor.cpp

```diff
@@ -8539,10 +8539,20 @@ void CHThorDiskReadBaseActivity::closepart()
                         dFile = &(super->querySubFile(subfile, true));
                     }
                 }
-                dFile->addAttrValue("@numDiskReads", curDiskReads);
                 StringBuffer clusterName;
                 dFile->getClusterName(0, clusterName);
+                cost_type previousReadCost = 0;
+                IPropertyTree & fileAttr = dFile->queryAttributes();
+                // Check if it is a legacy file without @readCost attribute -> calculate readCost and update file attributes (use previousReadCost to update file with full read costs )
+                if (!fileAttr.hasProp("@readCost") && fileAttr.hasProp("@numDiskReads"))
+                {
+                    stat_type prevDiskReads = fileAttr.getPropInt64("@numDiskReads", 0);
+                    previousReadCost = money2cost_type(calcFileAccessCost(clusterName, 0, prevDiskReads));
+                }
+
                 diskAccessCost = money2cost_type(calcFileAccessCost(clusterName, 0, curDiskReads));
+                dFile->addAttrValue("@numDiskReads", curDiskReads);
+                dFile->addAttrValue("@readCost", previousReadCost+diskAccessCost);
```

Comment:
That's correct.  The existing code will cause the file cost based on jhtree cache stats to be overwritten by the legacy code.  I've removed the code change to store the read & write cost (to avoid dynamic recalculation) from this PR and put in a separate one: https://track.hpccsystems.com/browse/HPCC-30911